### PR TITLE
change offset

### PR DESCRIPTION
**Pull request by takahashikzn**
Many tests are failed on JDK1.8.0_60.
It's seems that the constant pool layout have been changed.

### CODE REVIEW

**Comment by jhalterman** [comment]
Hi Takahishi - Thanks for bringing this to my attention. A few comments on your pull request:

* Do we know for sure if this change started with `1.8.0_60`? I hate depending on specific JRE releases.
* Another idea is that perhaps we can do an initial test of the constant pool at offset 2 to determine if that works with the user's JRE. We could do this in a static initializer against some publicly available lambda in the JDK (not sure if there are any?).
* Another idea is we set the offset based on the version, if the first attempt fails using that offset we can try another offset, if that fails we can set `SUPPORTS_LAMBDAS` to false.

**Comment by takahashikzn** [comment]
> Do we know for sure if this change started with 1.8.0_60? I hate depending on specific JRE releases.

I confirmed that this change was made with 1.8.0_51.

> Another idea is we set the offset based on the version, if the first attempt fails using that offset we can try another offset, if that fails we can set SUPPORTS_LAMBDAS to false.

I just updated my pull request. Can you confirm it?

File: src/main/java/net/jodah/typetools/TypeResolver.java [hunk 1]
```diff
@@ -62,9 +66,111 @@
...
+    final Class<?> returnType;
+
+    private LambdaTypeSample(Class<?> lambdaType, Class<?> returnType) {
+      this.lambdaType = lambdaType;
+      this.returnType = returnType;
+    }
+
+    static LambdaTypeSample get() throws ReflectiveOperationException {
+      Class<?> sliceOpsCls = Class.forName("java.util.stream.SliceOps");
+      Method castingArray = sliceOpsCls.getDeclaredMethod("castingArray");
```
Comment (jhalterman) [inline src/main/java/net/jodah/typetools/TypeResolver.java]: I'm a bit weary of depending on internal APIs here, which can change. What do you think of calculating the method ref offset the first time a lambda is resolved?

File: src/main/java/net/jodah/typetools/TypeResolver.java [hunk 2]
```diff
@@ -62,9 +66,111 @@
...
+  private static class AutoboxingMethodRefLambdaTypeSample {
+    final Class<?> lambdaType;
+
+    private AutoboxingMethodRefLambdaTypeSample(Class<?> lambdaType) {
+      this.lambdaType = lambdaType;
+    }
+
+    static AutoboxingMethodRefLambdaTypeSample get() throws ReflectiveOperationException {
+      Object collector =  Collectors.counting();
+      Field accumField = collector.getClass().getDeclaredField("accumulator");
```
Comment (jhalterman) [inline src/main/java/net/jodah/typetools/TypeResolver.java]: Same here - what do you think about doing this on demand the first time a user calls TypeTools?

**Comment by takahashikzn** [comment]
I could not think of calculating the offsets for the first time user calls TypeTools.
My pull request is resived.

**Comment by jroper** [comment]
I'd love to see this merged and a release with the fix, it's causing us problems in Play Framework.

**Comment by jhalterman** [comment]
@takahashikzn I like this new approach - good stuff! Two things before I merge:

* Do you think the existing tests cover this or are there any you can think to add?
* Can you squash the commits into a single commit?

@jroper I'll kick out a release immediately after this merges...

**Comment by jroper** [comment]
@jhalterman thanks!

**Comment by takahashikzn** [comment]
@jhalterman

> Do you think the existing tests cover this or are there any you can think to add?
I think the existing tests cover my modification because my mods are 'patch' and no method signature is changed.

> Can you squash the commits into a single commit?
I try squeezing my commits but there would be something wrong because I'm not used to using git.

**Comment by jroper** [comment]
@takahashikzn We have detailed instructions on how to squash commits here:

https://playframework.com/documentation/2.4.x/WorkingWithGit

**Comment by jhalterman** [comment]
Nvm - I squashed the commits and merged this. Thanks!

**Comment by jhalterman** [comment]
`0.4.3` is released and should be in maven central shortly.

File: src/main/java/net/jodah/typetools/TypeResolver.java [hunk 3]
```diff
@@ -347,6 +356,58 @@ private static void populateLambdaArgs(Class<?> functionalInterface, final Class
     }
   }
 
+  private static final class MethodRefOffsetResolutionFailed extends RuntimeException {}
+
+  /**
+   * Resolves method ref offset.
+   */
+  private static int resolveMethodRefOffset(ConstantPool constantPool, Class<?> lambdaType) {
```
Comment (JoshRosen) [inline src/main/java/net/jodah/typetools/TypeResolver.java]: This method's `lambdaType` argument appears to be unused. Is this a bug?

File: src/main/java/net/jodah/typetools/TypeResolver.java [hunk 4]
```diff
@@ -347,6 +356,58 @@ private static void populateLambdaArgs(Class<?> functionalInterface, final Class
...
+   * Resolves method ref offset of method reference lambda type having autoboxed return type.
+   * We can't cache the result value because results are different depending on given {@code lambdaType}.
+   */
+  private static int resolveAutoboxedMethodRefOffset(ConstantPool constantPool, Class<?> lambdaType) {
+    int constantPoolSize = constantPool.getSize();
+
+    for (int i = resolveMethodRefOffset(constantPool, lambdaType) + 1; i < constantPoolSize; i++) {
+      try {
+        // ignore constructor
+        if (!constantPool.getMemberRefInfoAt(constantPoolSize - i)[1].equals("<init>"))
```
Comment (JoshRosen) [inline src/main/java/net/jodah/typetools/TypeResolver.java]: What is the intuition for wanting to ignore the constructor in the autoboxed method ref case but not in the other case? Shouldn't we always ignore the `<init>` method's member ref info? Forgive me if this is a naive question.

File: src/main/java/net/jodah/typetools/TypeResolver.java (hunk 3 above)
Comment (jhalterman) [inline src/main/java/net/jodah/typetools/TypeResolver.java]: Probably not a bug - just an artifact of a previous refactoring, I think from a pull request. Thanks for pointing this out.

**Pull request closed without merging**